typegraph/core/src/runtimes/prisma/relationship/discovery.rs

```diff
@@ -0,0 +1,303 @@
+// Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
+// SPDX-License-Identifier: MPL-2.0
+
+use std::collections::HashSet;
+
+use crate::global_store::with_store;
+use crate::runtimes::prisma::type_utils::{self, has_fkey};
+use crate::types::Type;
+use crate::wit::core::TypeId;
+use crate::{errors::Result, types::Struct};
+
+use super::RelationshipRegistry;
+use super::{Cardinality, Relationship, RelationshipModel};
+
+/// scan model `model` for relationships
+pub fn scan_model(model: &Struct, registry: &RelationshipRegistry) -> Result<Vec<Relationship>> {
+    with_store(|s| {
+        let mut result = Vec::new();
+        let mut self_relationships = HashSet::new();
+
+        for (prop, ty) in model.data.props.iter() {
+            let ty = s.resolve_proxy(*ty)?;
+            // TODO why would it be possible to have self relationship?
```

Comment:
one example can be a hierarchical model "human" where the field parent points at itself